phoenicis-javafx/src/main/java/org/phoenicis/javafx/controller/engines/EnginesController.java

```diff
@@ -121,26 +129,84 @@ public EnginesView getView() {
     private void populateView(RepositoryDTO repositoryDTO, Map<String, Engine> engines) {
         this.repositoryCache = repositoryDTO;
         this.enginesCache = engines;
-        Platform.runLater(() -> {
-            this.enginesView.showWait();
-            List<CategoryDTO> categoryDTOS = new ArrayList<>();
-            for (TypeDTO typeDTO : repositoryDTO.getTypes()) {
-                if (typeDTO.getId().equals("engines")) {
-                    categoryDTOS = typeDTO.getCategories();
-                }
-            }
-            setDefaultEngineIcons(categoryDTOS);
-            this.enginesView.populate(this.enginesManager.getAvailableEngines(categoryDTOS), engines);
+
+        // show a waiting screen until the engines are loaded
+        enginesView.showWait();
+
+        // fetch all categories consisting of engines that are contained in the repository
+        final List<CategoryDTO> categoryDTOS = repositoryDTO.getTypes().stream()
+                .filter(type -> type.getId().equals("engines"))
+                .flatMap(type -> type.getCategories().stream())
+                .collect(Collectors.toList());
+
+        // generate the necessary css for the engine categories
+        setDefaultEngineIcons(categoryDTOS);
+
+        // fetch the engine categories objects contained in the engine categories
+        final Queue<EngineCategoryDTO> engineCategories = new ArrayDeque<>(
+                enginesManager.getAvailableEngines(categoryDTOS));
+
+        // insert the missing engine subcategories into the engine categories
+        fetchEngineSubcategories(engineCategories, ImmutableMap.of(), subcategoryMap -> {
+            final List<EngineCategoryDTO> categories = subcategoryMap.entrySet().stream()
+                    .map(entry -> new EngineCategoryDTO.Builder(entry.getKey())
+                            .withSubCategories(entry.getValue())
+                            .build())
+                    .collect(Collectors.toList());
+
+            Platform.runLater(() -> {
+                // update the view
+                enginesView.populate(categories, engines);
+            });
         });
     }
 
     /**
-     * forces an update of the view
+     * Fetches all engine subcategories that belong to a given list of engine categories
+     *
+     * @param engineCategories The engine categories
+     * @param result The temporary transport variable
+     * @param callback A callback method, which is called after all engine subcategories have been fetched
+     */
+    private void fetchEngineSubcategories(Queue<EngineCategoryDTO> engineCategories,
```

Comment:
Why `Queue`? What's the advantage to use recursion here?

Reply:
About the `Queue`:
A queue seems like the best data structure for this use case to me, because it allows for the retrieval and removal of the first object at the same time.
Ok, you could now argue that this is also possible with other collection types like `ArrayList` through the use of `remove()`, but in my opinion `remove` has a different semantic meaning.

About the use of `recursion`:
Recursion seemed to me like the only solution to implement this functionality without rewriting `EnginesManager`.
The issue about all our manager classes (e.g. `EnginesManager` and `RepositoryManager`) is that they all work with callbacks.
This makes it quite hard to concatenate multiple method calls without calling the "next" method inside the callback of the "current" method call.
This cries for a recursive solution to me.
Still I think it would be better to provide a "flat" way to work with our manager classes, i.e. a way which doesn't require callback methods.

Reply:
Ok. Yes, we should think about the callbacks again.